spv/sync.go

```diff
@@ -0,0 +1,1420 @@
+					walletID, prevChain[len(prevChain)-1].Hash, bestChain[len(bestChain)-1].Hash, len(prevChain))
+				for _, n := range prevChain {
+					s.sidechains.AddBlockNode(n)
+				}
+			}
+			tip := bestChain[len(bestChain)-1]
+			if len(bestChain) == 1 {
+				log.Infof("[%d] Connected block %v, height %d", walletID, tip.Hash, tip.Header.Height)
+			} else {
+				log.Infof("[%d] Connected %d blocks, new tip %v, height %d, date %v",
+					walletID, len(bestChain), tip.Hash, tip.Header.Height, tip.Header.Timestamp)
+			}
+
+			s.fetchHeadersProgress(headers[len(headers)-1])
+
+			s.sidechainMu.Unlock()
+		}
+
+		// Generate new locators
+		s.locatorMu.Lock()
+		locators, err = lowestChainWallet.BlockLocators(ctx, nil)
+		if err != nil {
+			s.locatorMu.Unlock()
+			return err
+		}
+		s.currentLocators = locators
+		s.locatorGeneration++
+		s.locatorMu.Unlock()
+	}
+}
+
+func (s *Syncer) fetchMissingCFilters(ctx context.Context, rp *p2p.RemotePeer) error {
+	for walletID, w := range s.wallets {
+		s.fetchMissingCfiltersStart(walletID)
+		progress := make(chan wallet.MissingCFilterProgress, 1)
+		go w.FetchMissingCFiltersWithProgress(ctx, rp, progress)
+
+		for p := range progress {
+			if p.Err != nil {
+				return p.Err
+			}
+			s.fetchMissingCfiltersProgress(walletID, p.BlockHeightStart, p.BlockHeightEnd)
+		}
+		s.fetchMissingCfiltersFinished(walletID)
+	}
+	return nil
+}
+
+func (s *Syncer) startupSync(ctx context.Context, rp *p2p.RemotePeer) error {
+	log.Infof("Syncing %d wallets", len(s.wallets))
+
+	_, _, highestChainWallet := s.highestChainTip(ctx)
+
+	// Disconnect from the peer if their advertised block height is
+	// significantly behind the wallet's.
+	_, tipHeight := highestChainWallet.MainChainTip(ctx)
+	if rp.InitialHeight() < tipHeight-6 {
+		return errors.E("peer is not synced")
+	}
+
+	s.fetchMissingCFilters(ctx, rp)
+
+	// Fetch any unseen headers from the peer.
+	s.fetchHeadersStart(rp.InitialHeight())
+	log.Infof("About to fetch headers")
+	log.Debugf("Fetching headers from %v", rp.RemoteAddr())
+	err := s.getHeaders(ctx, rp)
+	if err != nil {
+		return err
+	}
+	s.fetchHeadersFinished()
+	log.Infof("Finished Fetching Headers")
+
+	if atomic.CompareAndSwapUint32(&s.atomicCatchUpTryLock, 0, 1) {
+		for key, w := range s.wallets {
+			err = func() error {
+				rescanPoint, err := w.RescanPoint(ctx)
+				if err != nil {
+					return err
+				}
+				if rescanPoint == nil {
+					if !s.loadedFilters[key] {
+						s.filterWalletID = key
+						err = w.LoadActiveDataFilters(ctx, s, true)
+						s.filterWalletID = -1
+						if err != nil {
+							return err
+						}
+						s.loadedFilters[key] = true
+					}
+
+					s.synced(key)
+
+					return nil
+				}
+				// RescanPoint is != nil so we are not synced to the peer and
+				// check to see if it was previously synced
+				s.unsynced(key)
+
+				s.discoverAddressesStart(key)
+				err = w.DiscoverActiveAddresses(ctx, rp, rescanPoint, !w.Locked() && s.discoverAccountsIfUnlocked)
```

Comment:
I believe we should always want to discover accounts for unlocked wallets, we need only check if a wallet is unlocked in order to perform account discovery. The `s.discoverAccountsIfUnlocked` variable can be removed.